File: DSA/Linkedlist/Code/DetectCycleInLinkedList.py
# Detect Cycle in Linked List in Python

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class CycleDetection:
    def __init__(self):
        self.head = None
    
    def add_node(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return
        temp = self.head
        while temp.next is not None:
            temp = temp.next
        temp.next = new_node
    
    # Floyd's Cycle Detection (Tortoise & Hare)
    def detect_cycle_floyd(self):
        if self.head is None:
            return False
        
        slow = self.head
        fast = self.head
        
        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next
            
            if slow == fast:
                return True
        
        return False
    
    # Hash Set approach
    def detect_cycle_hash_set(self):
        visited = set()
        temp = self.head
        
        while temp is not None:
            if id(temp) in visited:
                return True
            visited.add(id(temp))
            temp = temp.next
        
        return False
    
    # Find cycle start node
    def find_cycle_start(self):
        if self.head is None:
            return None
        
        slow = self.head
        fast = self.head
        
        # Detect cycle
        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next
            
            if slow == fast:
                break
        
        if fast is None or fast.next is None:
            return None
        
        # Find cycle start
        slow = self.head
        while slow != fast:
            slow = slow.next
            fast = fast.next
        
        return slow

# Test
print("=== Detect Cycle in Linked List ===\n")

# Test 1: No cycle
list1 = CycleDetection()
list1.add_node(1)
list1.add_node(2)
list1.add_node(3)
list1.add_node(4)

print("List 1 (No cycle):", "Cycle detected" if list1.detect_cycle_floyd() else "No cycle")

# Test 2: With cycle
list2 = CycleDetection()
n1 = Node(1)
n2 = Node(2)
n3 = Node(3)
n4 = Node(4)

n1.next = n2
n2.next = n3
n3.next = n4
n4.next = n2  # Creates cycle

list2.head = n1

print("List 2 (With cycle):", "Cycle detected" if list2.detect_cycle_floyd() else "No cycle")

cycle_start = list2.find_cycle_start()
if cycle_start is not None:
    print(f"Cycle starts at node with data: {cycle_start.data}")

print("\nComplexity Analysis:")
print("Floyd's Cycle Detection: O(n) time, O(1) space")
print("Hash Set approach: O(n) time, O(n) space")
print("Find Cycle Start: O(n) time, O(1) space")

File: DSA/Linkedlist/Code/DetectCycleInLinkedList.java
// Detect Cycle in Linked List in Java

import java.util.HashSet;

class Node {
    int data;
    Node next;
    
    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

public class CycleDetection {
    Node head;
    
    void addNode(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            return;
        }
        Node temp = head;
        while (temp.next != null) temp = temp.next;
        temp.next = newNode;
    }
    
    // Floyd's Cycle Detection
    boolean detectCycleFloyd() {
        if (head == null) return false;
        
        Node slow = head;
        Node fast = head;
        
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            
            if (slow == fast)
                return true;
        }
        
        return false;
    }
    
    // Hash Set approach
    boolean detectCycleHashSet() {
        HashSet<Node> visited = new HashSet<>();
        Node temp = head;
        
        while (temp != null) {
            if (visited.contains(temp))
                return true;
            visited.add(temp);
            temp = temp.next;
        }
        
        return false;
    }
    
    // Find cycle start node
    Node findCycleStart() {
        if (head == null) return null;
        
        Node slow = head;
        Node fast = head;
        
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            
            if (slow == fast)
                break;
        }
        
        if (fast == null || fast.next == null)
            return null;
        
        slow = head;
        while (slow != fast) {
            slow = slow.next;
            fast = fast.next;
        }
        
        return slow;
    }
    
    public static void main(String[] args) {
        CycleDetection list = new CycleDetection();
        
        System.out.println("=== Detect Cycle in Linked List ===\n");
        
        // Test 1: No cycle
        list.addNode(1);
        list.addNode(2);
        list.addNode(3);
        list.addNode(4);
        
        System.out.print("List 1 (No cycle): ");
        System.out.println(list.detectCycleFloyd() ? "Cycle detected" : "No cycle");
        
        // Test 2: With cycle (manual)
        Node head = new Node(1);
        Node n2 = new Node(2);
        Node n3 = new Node(3);
        Node n4 = new Node(4);
        
        head.next = n2;
        n2.next = n3;
        n3.next = n4;
        n4.next = n2; // Creates cycle
        
        System.out.print("List 2 (With cycle): ");
        
        // Detect using Floyd's algorithm
        Node slow = head;
        Node fast = head;
        boolean hasCycle = false;
        
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            
            if (slow == fast) {
                hasCycle = true;
                break;
            }
        }
        
        System.out.println(hasCycle ? "Cycle detected" : "No cycle");
        
        System.out.println("\nComplexity Analysis:");
        System.out.println("Floyd's Cycle Detection: O(n) time, O(1) space");
        System.out.println("Hash Set approach: O(n) time, O(n) space");
        System.out.println("Find Cycle Start: O(n) time, O(1) space");
    }
}
